patches/tModLoader/Terraria/GameContent/TeleportPylonsSystem.cs.patch

```diff
@@ -1,11 +1,209 @@
++						Tile tile = Framing.GetTileSafely(value.Position);
++						ModPylon pylon = TileLoader.GetTile(tile.type) as ModPylon;
++						teleportPylonInfo.ModName = pylon!.Mod.Name;
++						teleportPylonInfo.ModPylonName = pylon!.Name;
++					}
+ 					TeleportPylonInfo item = teleportPylonInfo;
+-					_pylons.Add(item);
++					AddNewPylon(item);
+ 				}
+ 			}
+ 
+@@ -67,12 +_,14 @@
+ 		}
+ 
+ 		public void AddForClient(TeleportPylonInfo info) {
+-			if (!_pylons.Contains(info))
+-				_pylons.Add(info);
++			if (!_pylons.Contains(info)) 
++				AddNewPylon(info);
+ 		}
+ 
+ 		public void RemoveForClient(TeleportPylonInfo info) {
+-			_pylons.RemoveAll((TeleportPylonInfo x) => x.Equals(info));
++			_pylons.RemoveAll(x => x.Equals(info));
++			_vanillaPylons.RemoveAll(x => x.Equals(info));
++			_moddedPylons.RemoveAll(x => x.Equals(info));
+ 		}
+ 
+ 		public void HandleTeleportRequest(TeleportPylonInfo info, int playerIndex) {
+@@ -93,7 +_,14 @@
+ 			}
+ 
+ 			if (flag) {
+-				flag &= !NPC.AnyDanger(quickBossNPCCheck: false, ignorePillars: true);
++				bool forceSucceed = false;
++				if (PylonLoader.ValidTeleportCheck_PreAnyDanger(info) is { } value) {
++					forceSucceed = value;
++					flag &= value;
++				}
++				flag &= (info.TypeOfPylon == TeleportPylonType.Modded
++					? ModContent.TryFind(info.ModName, info.ModPylonName, out ModPylon pylon) && pylon.ValidTeleportCheck_AnyDanger(info)
++					: !NPC.AnyDanger(quickBossNPCCheck: false, ignorePillars: true)) || forceSucceed;
+ 				if (!flag)
+ 					key = "Net.CannotTeleportToPylonBecauseThereIsDanger";
+ 			}
+@@ -120,9 +_,11 @@
+ 				if (!flag)
+ 					key = "Net.CannotTeleportToPylonBecauseNotMeetingBiomeRequirements";
+ 			}
++			if (info.TypeOfPylon == TeleportPylonType.Modded && ModContent.TryFind(info.ModName, info.ModPylonName, out ModPylon destinationPylon))
++				destinationPylon.ValidTeleportCheck_DestinationPostCheck(info, ref flag, ref key);
+ 
+-			if (flag) {
+-				bool flag2 = false;
++			TeleportPylonInfo nearbyInfo = new TeleportPylonInfo();
++			bool flag2 = false;
+ 				int num = 0;
+ 				for (int i = 0; i < _pylons.Count; i++) {
+ 					TeleportPylonInfo info2 = _pylons[i];
+@@ -147,13 +_,13 @@
+ 						sceneMetrics2.ScanAndExportToMain(settings);
+ 						if (DoesPylonAcceptTeleportation(info2, player)) {
+ 							flag2 = true;
++							nearbyInfo = info2;
+ 							break;
+ 						}
+ 					}
+ 				}
+ 
+ 				if (!flag2) {
+-					flag = false;
+ 					switch (num) {
+ 						default:
+ 							key = "Net.CannotTeleportToPylonBecausePlayerIsNotNearAPylon";
+@@ -166,9 +_,11 @@
+ 							break;
+ 					}
+ 				}
+-			}
++			if (info.TypeOfPylon == TeleportPylonType.Modded && ModContent.TryFind(nearbyInfo.ModName, nearbyInfo.ModPylonName, out ModPylon nearbyPylon))
++				nearbyPylon.ValidTeleportCheck_NearbyPostCheck(nearbyInfo, ref flag, ref key);
++			PylonLoader.PostValidTeleportCheck(info, nearbyInfo, ref flag, ref flag2, ref key);
+ 
+-			if (flag) {
++			if (flag && flag2) {
```

Comment:
in comparison, no comment required here. Clear from above proposed comments, and the fact this change is additive what it does